pdfparser/poppler.pyx

```diff
@@ -575,134 +577,211 @@ cdef class LinesIterator:
+    property text:
+        def __get__(self):
+            if self.text is None:
+                self._cache_line_data()
+            return self.text
 
-            # gets bounding boxes for all characters and font info
-            for i in range(word_length):
-                word.getCharBBox(i, &bx1, &by1, &bx2, &by2)
+    property char_bboxes:
+        def __get__(self):
+            if self.char_bboxes is None:
+                self._cache_line_data()
+            return self.char_bboxes
 
-                last_bbox = BBox(bx1, by1, bx2, by2)
+    property char_fonts:
+        def __get__(self):
+            if self.char_fonts is None:
+                self._cache_line_data()
+            return self.char_fonts
 
-                # if previous word is space update it's right end
-                if i == 0 and words and words[-1] == u' ':
-                    self._bboxes[-1].x2 = last_bbox.x1
 
-                self._bboxes.append(last_bbox)
+cdef class WordsIterator:
+    cdef PopplerPage page
+    cdef TextWord *words
 
-                word.getColor(&r, &g, &b)
+    def __cinit__(self, PopplerPage page, Line line):
+        self.page = page
+        self.words = line.getWords()
 
-                _font_name = word.getFontName(i)
+    def __next__(self):
+        cdef Word word
 
-                # If the current character comes out of NOT
-                # embedded font `_font_name` is NULL.
-                if _font_name != NULL:
-                    try:
-                        font_name = _font_name.getCString().decode('UTF-8')
-                    except:
-                        font_name = u'%r' % _font_name.getCString()
-                else:
-                    font_name = u'?'
+        if not self.words:
+            raise StopIteration()
 
-                last_font = FontInfo(font_name,
-                                     word.getFontSize(),
-                                     Color(r, g, b))
+        word = Word(self.page).wrap(self.words)
+        self.words = self.words.getNext()
+        return word
 
-                self._fonts.append(last_font)
 
-            _word_text = word.getText()
-            words.append(_word_text.getCString().decode('UTF-8'))
-            del _word_text
+cdef class Word:
+    cdef:
+        PopplerPage  page
+        TextWord    *word
+        unicode      text
+        BBox         bbox
+        list         char_bboxes
+        CompactList  char_fonts
+        bool         has_space_after
 
-            # the number of bboxes/fonts must equal the number of characters in the word
-            assert len(words[-1]) == word_length
+    def __cinit__(self, PopplerPage page):
+        self.page = page
+        self.word = NULL
+
+    cdef Word wrap(self, TextWord *word):
+        cdef:
+            GooString *_text
+            GooString *_font_name
+            int        i, word_length
+            double     l, t, r, b, g
```

Comment:
Possibly change to `left, top, right, bottom, red, green, blue` or at least comment that `r` is reused for `right`/`red` and `b` for `bottom`/`blue` 🎃